src/Agent.cs

```diff
@@ -0,0 +1,237 @@
+    /// <param name="action">The action being measured</param>
+    /// <returns>The result of executing the action.</returns>
+    public T TimeMs<T>(String metricName, Func<T> action)
+    {
+      return ActuallyTime(metricName, action, 1000);
+    }
+
+    /// <summary>
+    ///   Measure the duration of an action in milliseconds.
+    /// </summary>
+    /// <param name="metricName">The name of the metric to send.  Letters, numbers, and dots.</param>
+    /// <param name="action">The action being measured</param>
+    public void TimeMs(String metricName, Action action)
+    {
+      ActuallyTime(metricName, action, 1000);
+    }
+
+    private T ActuallyTime<T>(String metricName, Func<T> action, float durationMultiplier = 1)
+    {
+      var start = DateTime.Now;
+      try
+        {
+          return action();
+        }
+      finally
+        {
+          var end = DateTime.Now;
+          var duration = end - start;
+          Gauge(metricName, (float)duration.TotalSeconds * durationMultiplier);
+        }
+    }
+
+    private void ActuallyTime(String metricName, Action action, float durationMultiplier = 1)
+    {
+      var start = DateTime.Now;
+      try
+        {
+          action();
+        }
+      finally
+        {
+          var end = DateTime.Now;
+          var duration = end - start;
+          Gauge(metricName, (float)duration.TotalSeconds * durationMultiplier);
+        }
+    }
+
+    /// <summary>
+    ///   Measure an occurance of something in your code.
+    /// </summary>
+    /// <param name="metricName">The name of the metric to send.  Letters, numbers, and dots.</param>
+    /// <param name="value">The value to record.</param>
+    /// <param name="time">The exact time at which the event happened.  You probably want the default.</param>
+    /// <param name="count">The number of events which this represents.  You almost certainly want the default.</param>
+    /// <returns>The value being passed in via value, or null if something bad happened.</returns>
+    public float? Increment(String metricName, float value = 1, DateTime? time = null, int count = 1)
+    {
+      try
+        {
+          if(!ValidateMetricName(metricName)) return null;
+          if (!Enabled) return value;
+          int metricTime = (time ?? DateTime.Now).ToEpoch();
+          _collector.SendMessage(String.Format("increment {0} {1} {2} {3}", metricName, value, metricTime, count));
+          return value;
+        }
+      catch (Exception e)
+        {
+          ReportException(e);
+        }
+      return null;
+    }
+
+    /// <summary>
+    ///   Tag a point or duration in time with a note, like a deploy or a service outage.
+    /// </summary>
+    /// <param name="message">The note you want to show up in your metrics.  A sentence or less, no newlines.</param>
+    /// <param name="time">The exact time at which the event happened.  You probably want the default.</param>
```

Comment:
After reading a few of these parameter comments I wondered if we should say that you should pass in a `DateTime`, but I guess in csharp your IDE will be telling you the type you want, yeah?